File: Algorithms/p378_Kth_Smallest_Element_in_a_Sorted_Matrix_1.py
class Solution(object):
    def kthSmallest(self, matrix, k):
        """
        :type matrix: List[List[int]]
        :type k: int
        :rtype: int
        """
        m_len = len(matrix)
        for i in range(0,k-1):
            matrix[0][0] = None
            j1 = 0
            j2 = 0
            while j1<m_len and j2<m_len:
                v1 = matrix[j1+1][j2] if j1!=m_len-1 else None
                v2 = matrix[j1][j2+1] if j2!=m_len-1 else None
                if v1!=None and (v2==None or v1<=v2):
                    matrix[j1][j2],matrix[j1+1][j2] = v1,None
                    j1 += 1
                elif v2!=None and (v1==None or v2<=v1):
                    matrix[j1][j2],matrix[j1][j2+1] = v2,None
                    j2 += 1
                else:
                    break
        return matrix[0][0]

File: Algorithms/p378_Kth_Smallest_Element_in_a_Sorted_Matrix/p378_Kth_Smallest_Element_in_a_Sorted_Matrix_1.cpp
class Solution {
public:
    int kthSmallest(vector<vector<int>>& matrix, int k) {
        const int max = 0x7fffffff;
        int len = matrix.size();
        for(int i=0;i<k-1;i++){
            matrix[0][0] = max;
            for(int j1=0,j2=0;j1<len&&j2<len;){
                int v1 = matrix[j1][j2];
                int v2 = j1!=len-1 ? matrix[j1+1][j2] : max;
                int v3 = j2!=len-1 ? matrix[j1][j2+1] : max;
                if(v2<v1 && v2<=v3){
                    matrix[j1][j2] = v2;
                    matrix[j1+1][j2] = v1;
                    j1++;
                }else if(v3<v1 && v3<=v2){
                    matrix[j1][j2] = v3;
                    matrix[j1][j2+1] = v1;
                    j2++;
                }else{
                    break;
                }
            }
        }
        return matrix[0][0];
    }
};
